src/sage/combinat/partition.py

```diff
@@ -8356,6 +8405,24 @@ class Partitions_with_constraints(IntegerListsLex):
     Element = Partition
     options = Partitions.options
 
+    def __contains__(self, x):
+        """
+        Check if ``x`` is contained in ``self``.
+
+        TESTS::
+
+            sage: P = Partitions(4, max_slope=2)
+            sage: [3,1] in P
+            True
+            sage: [3,1,0] in P
+            True
+        """
+        if x in _Partitions:
+            for i in range(len(x)-1, -1, -1):
+                if x[i]:
+                    x = x[:i+1]
+                    break
+            return super().__contains__(x)
```

Comment:
```suggestion
        # strip off trailing 0s
        for i in range(len(x)-1, -1, -1):
            if x[i] != 0:
                x = x[:i+1]
                break
        else:
            x = []
        return super().__contains__(x)
```
Before, this could return `None` (which is an error IMO). However, I think it would be double checking the partition requirements. Note that we cannot simply check `if not x[i]:` since it could be `None`. Please also add a doctest for such "weird" cases.

Reply:
Unfortunately, this doesn't quite work, because `IntegerListsLex.__contains__` fails if the list is not a list of integers:
```
sage: [2.5,0.5] in IntegerListsLex(3)
True
sage: [None] in IntegerListsLex(3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[19], line 1
----> 1 [None] in IntegerListsLex(Integer(3))

File ~/sage/src/sage/combinat/integer_lists/lists.py:279, in IntegerLists.__contains__(self, item)
    268 def __contains__(self, item):
    269     """
    270     Return ``True`` if ``item`` meets the constraints imposed by
    271     the arguments.
   (...)
    277         True
    278     """
--> 279     return self.backend._contains(item)

File ~/sage/src/sage/combinat/integer_lists/base.pyx:208, in sage.combinat.integer_lists.base.IntegerListsBackend._contains()
    206 if len(comp) < self.min_length or len(comp) > self.max_length:
    207     return False
--> 208 n = sum(comp)
    209 if n < self.min_sum or n > self.max_sum:
    210     return False

TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'
```
I agree that this should be fixed.

I was a bit puzzled that `super().__contains__(x)` checks that `x` is actually a partition.  This is because we are silently passing `max_slope=0`.

Curiously, my version does *not* return `None`, it seems that `__contains__` is a bit magic.